pkg/sql/sem/tree/cast.go

```diff
@@ -1253,24 +1255,70 @@ func ValidCast(src, tgt *types.T, ctx CastContext) bool {
 
 	// If src and tgt are not both array or tuple types, check castMap for a
 	// valid cast.
-	c, ok := lookupCast(
-		src.Oid(),
-		tgt.Oid(),
-		false, /* intervalStyleEnabled */
-		false, /* dateStyleEnabled */
-	)
+	c, ok := lookupCast(src, tgt, false /* intervalStyleEnabled */, false /* dateStyleEnabled */)
 	if ok {
 		return c.maxContext >= ctx
 	}
 
 	return false
 }
 
-// lookupCast returns a cast that describes the cast from src to tgt if
-// it exists. If it does not exist, ok=false is returned.
-func lookupCast(src, tgt oid.Oid, intervalStyleEnabled, dateStyleEnabled bool) (cast, bool) {
-	if tgts, ok := castMap[src]; ok {
-		if c, ok := tgts[tgt]; ok {
+// lookupCast returns a cast that describes the cast from src to tgt if it
+// exists. If it does not exist, ok=false is returned.
+func lookupCast(src, tgt *types.T, intervalStyleEnabled, dateStyleEnabled bool) (cast, bool) {
+	srcFamily := src.Family()
+	tgtFamily := tgt.Family()
+	srcFamily.Name()
+
+	// Unknown is the type given to an expression that statically evaluates
+	// to NULL. NULL can be immutably cast to any type in any context.
+	if srcFamily == types.UnknownFamily {
+		return cast{
+			maxContext: CastContextImplicit,
+			volatility: VolatilityImmutable,
+		}, true
+	}
+
+	// Enums have dynamic OIDs, so they can't be populated in castMap. Instead,
+	// we dynamically create cast structs for valid enum casts.
+	// lookups for valid enum casts
+	if srcFamily == types.StringFamily && tgtFamily == types.EnumFamily {
+		// Casts from string types to enums are immutable and allowed in
+		// explicit contexts.
+		return cast{
+			maxContext: CastContextExplicit,
+			volatility: VolatilityImmutable,
```

Comment:
i would've thought enums were stable, considering they require an evaluation of the current TypeDescriptor to determine whether or not they are valid.

also wonder if this is more readable as (feel free to ignore)

```
if destFamily is enum {
   switch srcFamily {
      case StringFamily:
          return cast { ... }      
   }
}
```

Reply:
> also wonder if this is more readable as (feel free to ignore)

👍 done

> i would've thought enums were stable

You're absolutely right. I copied the current volatilities which are incorrect, causing correctness bugs. I opened #74470 with an example, and I've added a commit to this PR to fix it. I think that commit should be backported, so I'll do that unless you disagree.

Reply:
backporting requires a discussion on compatibility, but that commit LGTM

Reply:
(in fact, it'd be similar to the work ebony is doing in that we'll have to block or automatically update statements that are violating this)

Reply:
I'm not sure what you mean. Any table expressions with one of these casts is invalid. If we backport this as-is to a cluster that already has a table expression with one of these casts, it shouldn't change anything besides preventing new table expressions with the cast from being created, right?